16-reinforce/bilstm-tagger.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+
+#the parameters from mixer
+NXENT = 40
+NXER = 20
+delta = 2
+
+# format of files: each line is "word1|tag1 word2|tag2 ..."
+train_file = "../data/tags/train.txt"
+dev_file = "../data/tags/dev.txt"
+
+w2i = defaultdict(lambda: len(w2i))
+t2i = defaultdict(lambda: len(t2i))
+
+
+def read(fname):
+    """
+    Read tagged file
+    """
+    with open(fname, "r") as f:
+        for line in f:
+            words, tags = [], []
+            for wt in line.strip().split():
+                w, t = wt.split('|')
+                words.append(w2i[w])
+                tags.append(t2i[t])
+            yield (words, tags)
+
+
+# Read the data
+train = list(read(train_file))
+unk_word = w2i["<unk>"]
+w2i = defaultdict(lambda: unk_word, w2i)
+unk_tag = t2i["<unk>"]
+t2i = defaultdict(lambda: unk_tag, t2i)
+nwords = len(w2i)
+ntags = len(t2i)
+dev = list(read(dev_file))
+
+# DyNet Starts
+model = dy.Model()
+trainer = dy.AdamTrainer(model)
+
+# Model parameters
+EMBED_SIZE = 64
+HIDDEN_SIZE = 128
+
+# Lookup parameters for word embeddings
+LOOKUP = model.add_lookup_parameters((nwords, EMBED_SIZE))
+
+# Word-level BiLSTM
+LSTM = dy.BiRNNBuilder(1, EMBED_SIZE, HIDDEN_SIZE, model, dy.LSTMBuilder)
+
+# Word-level softmax
+W_sm = model.add_parameters((ntags, HIDDEN_SIZE))
+b_sm = model.add_parameters(ntags)
+
+
+#Baseline reward parameters
+W_bl_p = model.add_parameters((1, HIDDEN_SIZE))
+b_bl_p = model.add_parameters(1)
+
+# Calculate the scores for one example
+def calc_scores(words):
+    dy.renew_cg()
+
+    # Transduce all batch elements with an LSTM
+    word_reps = LSTM.transduce([LOOKUP[x] for x in words])
+
+    # Softmax scores
+    W = dy.parameter(W_sm)
+    b = dy.parameter(b_sm)
+    scores = [dy.affine_transform([b, W, x]) for x in word_reps]
+
+    return scores
+
+
+# Calculate MLE loss for one example
+def calc_loss(scores, tags):
+    losses = [dy.pickneglogsoftmax(score, tag) for score, tag in zip(scores, tags)]
+    return dy.esum(losses)
+
+def calc_reinforce_loss(words, tags, delta):
+    dy.renew_cg()
+
+    # Transduce all batch elements with an LSTM
+    word_reps = LSTM.transduce([LOOKUP[x] for x in words])
+
+    # Softmax scores
+    W = dy.parameter(W_sm)
+    b = dy.parameter(b_sm)
+
+    #calculate the probability distribution 
+    scores = [dy.affine_transform([b, W, x]) for x in word_reps]
+    losses = [dy.pickneglogsoftmax(score, tag) for score, tag in zip(scores, tags)]
+    probs = [-dy.exp(loss).as_array() for loss in losses]
+
+    #then take samples from the probability distribution
+    samples = [np.random.choice(range(len(x)), p=x) for x in probs]
+
+    #calculate accuracy=reward
+    correct = [sample == tag for sample, tag in zip(samples, tags)]
+    r_i = float(sum(correct))/len(correct)
+    r = dy.constant((1), r_i)
+    # Reward baseline for each word
+    W_bl = dy.parameter(W_bl_p)
+    b_bl = dy.parameter(b_bl_p)
+    r_b = [dy.affine_transform([b_bl, W_bl, x]) for x in word_reps]
```

Comment:
I think you actually need to break the computation graph here by wrapping `x` in `dy.nobackprop()`

Reply:
Good call!  I was using pretrained word representations in my c++ code and completely missed that here.